integration-tests/deployment/environment.go

```diff
@@ -35,22 +39,43 @@ type OffchainClient interface {
 	jobv1.JobServiceClient
 	nodev1.NodeServiceClient
 	csav1.CSAServiceClient
+	// sometimes we need to connect to the nodes directly to execute actions that are not supported by JD/FMS
+	NodeClients() []*client.ChainlinkK8sClient
 }
 
+// TODO: we should rename it EVM chain, as it's not generic at all
 type Chain struct {
 	// Selectors used as canonical chain identifier.
 	Selector uint64
 	Client   OnchainClient
 	// Note the Sign function can be abstract supporting a variety of key storage mechanisms (e.g. KMS etc).
-	DeployerKey *bind.TransactOpts
+	// Rename to something more universal, as these keys are not only for deploying, but also for other actions
+	Keys []*bind.TransactOpts
+	// Function to execute if transaction submission fails.
+	RetrySubmit func(tx *types.Transaction, err error) (*types.Transaction, error)
 	Confirm     func(tx common.Hash) (uint64, error)
+	DefaultKey  func() *bind.TransactOpts
+}
+
+// NoOpRetrySubmit is a retry submit function that does nothing.
+func NoOpRetrySubmit(_ *types.Transaction, err error) (*types.Transaction, error) {
+	return nil, err
+}
+
+// we do need mocks for higher-level environments
```

Comment:
this also breaks the env abstraction - can you elaborate on their use case? Since they are unused in this PR maybe drop for now and we can discuss them in a follow up